### ISSUE TITLE
Improve performance of JxW calculations

### ISSUE THREAD

**Issue by drwells** [IBAMR/IBAMR#444]
libMesh implements this (for isoparametric elements) in a very general way that is lacking some optimizations that might help us. We spend about 7% of our overall run time just computing `JxW` (that is, Jacobian times quadrature weight) because loops are in different functions, branches are in the wrong order (e.g., at each quadrature point we check what computations we need to do, instead of checking that outside the quadrature point loop and then doing the loop over exactly what we need), and the data arrays are not stored contiguously.

libMesh does have special code for simplex elements since they can use affine maps; our performance woes mostly come from hexahedral meshes where the Jacobian varies from point to point.

If we want to fix this we will have to roll our own Jacobian calculation code that supplants what `FEMap` does. Some ideas:
1. dimension-specific code: it would be better to completely evade libMesh's dimension-independent code and use known dimensionality to unroll loops.
2. geometry-specific code: if we are calculating Jacobians on a hexahedron then we know that there are exactly eight vertices.
3. Take the data out of `FEMap` for reference values of gradients and store them in a more optimal data structure than `std::vector<std::vector<double>>`.

For example, here is the deal.II code to compute the contravariant transformation (see `source/fe/mapping_q_generic.cc`, implementation of `maybe_update_Jacobians`):
```cpp
              for (unsigned int point = 0; point < n_q_points; ++point)
                {
                  const Tensor<1, dim> *data_derv =
                    &data.derivative(point + data_set, 0);

                  double result[spacedim][dim];

                  // peel away part of sum to avoid zeroing the
                  // entries and adding for the first time
                  for (unsigned int i = 0; i < spacedim; ++i)
                    for (unsigned int j = 0; j < dim; ++j)
                      result[i][j] = data_derv[0][j] * supp_pts[0][i];
                  for (unsigned int k = 1; k < data.n_shape_functions; ++k)
                    for (unsigned int i = 0; i < spacedim; ++i)
                      for (unsigned int j = 0; j < dim; ++j)
                        result[i][j] += data_derv[k][j] * supp_pts[k][i];

                  // write result into contravariant data. for
                  // j=dim in the case dim<spacedim, there will
                  // never be any nonzero data that arrives in
                  // here, so it is ok anyway because it was
                  // initialized to zero at the initialization
                  for (unsigned int i = 0; i < spacedim; ++i)
                    for (unsigned int j = 0; j < dim; ++j)
                      data.contravariant[point][i][j] = result[i][j];
                }
```
The innermost loops are candidates for unrolling since `dim` and `spacedim` are template parameters; we traverse two contiguous data arrays (`supp_pts` and `data_derv`) to compute the result.

**Comment by drwells** [comment]
@boyceg what would you think of assigning this as a project for a graduate student? I can handle the code review process. It would be a nice introduction to 'how to write fast unstructured grid FE codes).

**Comment by drwells** [comment]
I have a few more thoughts on this. I am pretty sure I know how to do this in an efficient way.

We should come up with an abstract base class `JacobianBase` that has a single virtual method `const std::vector<double>* JacobianBase::get_jacobians(const Elem *elem)`. We then inherit from `JacobianBase` with classes that have fixed (via template parameters) dimension, space dimension, and element shape. For non-affine maps these classes will store, in fixed-size arrays, all relevant data computed on the reference element as well as the correct reference element quadrature points (all this stuff is already available in `FEMap`, but we can do better if we copy the arrays into our own data structures based on `std::array`). Then each inheriting class can just read off the coordinates of the element nodes (again into its own data structures) and then fill out the array of Jacobians.

Finally, we can pull down the correct class inheriting from `JacobianBase` by copying and pasting `FEMapCache` one more time: each quadrature rule corresponds to exactly one set of reference data and element shape. We should be able to set all loop bounds except for the total number of quadrature points at compile time (we could do that too but its not necessary, particularly if there are inner loops).

**Comment by drwells** [comment]
@jabrown893 and I sat down to chat about this today. The code is available on

https://github.com/IBAMR/IBAMR/tree/drwells/jacobian-calculator

**Comment by drwells** [comment]
@jabrown893 I added some more code to the branch to implement a few things we discussed today.

I think that we may as well just skip all usage of `libMesh::FEMap` and just implement our own lookup tables: the only elements that need it are `Tri6`, `Tet10`, and `Hex27` so it shouldn't be a ton of work to compute these gradients ourselves.

**Comment by jabrown893** [comment]
@drwells I added a draft of the QuadrilateralJacobianCalculator, if you have a chance to take a look.

**Comment by drwells** [comment]
@jabrown893 I would like to wrap this up soon. Do you have time to work on this this week?

**Comment by drwells** [comment]
@jabrown893 *ping*

**Comment by drwells** [comment]
@jabrown893 This is still a big bottleneck (about 10% of total compute time) in the heart model and I really want this to be finished as soon as possible. Lets set up a weekly time to work on this until its done :)

**Comment by drwells** [comment]
@jabrown893 I finished the implementation of `Quad9` from last week: the code is the 'right' way to implement this since we only have to calculate Jacobians at the underlying 1D quadrature points. I verified that we get a fourth order accurate approximation to the area of a circle.

We should do the same thing for `Hex8` and `Hex27`. It might make sense to move some of the code out of the implementation of `Quad9` so that we can share it with the other two. Unfortunately `Tet10` is going to be much messier.

**Comment by drwells** [comment]
I also finished a generic implementation, which is the best we can do for `Tet10` and `Tet6`.

**Comment by drwells** [comment]
@jabrown893 I want to finish this (at least the completely generic implementation) and use it in FEDataManager by the end of next week. Will you have time to work on it?

**Comment by jabrown893** [comment]
@drwells I have to grade exams at the beginning of the week, but I may have time to work on it toward the very end of the week. I'm not sure if that will work with your timeline, though.

**Comment by drwells** [comment]
I'd like you to take a stab at writing a class like `FEMapCache` - it should only take an hour or two. Can you find a block next Thursday or Friday for this?

**Comment by jabrown893** [comment]
@drwells Okay, I just pushed the commit. I think all that is left to do is to write tests to see if FEDataManager implements it correctly?

**Comment by drwells** [comment]
I think that we are okay as long as the current tests using FEDataManager pass - if we don't impact the values of spreading or interpolation we should be good. If we want to test `JacobianCalculatorCache` directly then we can just add some more code to the test provided on the branch.

Would you please open a PR for the branch now that its mostly done?

**Comment by drwells** [comment]
Fixed by #764.
